catboost/private/libs/algo/nonsymmetric_index_calcer.cpp

```diff
@@ -214,6 +213,124 @@ void UpdateIndices(
     );
 }
 
+void UpdateIndicesWithSplit(
+    const TSplitNode& node,
+    const TTrainingDataProviders& trainingData,
+    const TSimpleSpan<ui32>& docsSubset,
+    const TFold& fold,
+    NPar::TLocalExecutor* localExecutor,
+    TArrayRef<TIndexType> indicesRef, TSimpleSpan<ui32>& l, TSimpleSpan<ui32>& r
+) {
+    const ui32 doc_size = docsSubset.size();
+    const ui32* docsSubsetPtr = docsSubset.data();
+    TQuantizedObjectsDataProviderPtr objectsDataProvider;
+    const ui32* columnsIndexing;
+    TIndexedSubsetCache indexedSubsetCache; // not really updated because it is used for test only
+    GetObjectsDataAndIndexing(
+        trainingData,
+        fold,
+        node.Split.Type == ESplitType::EstimatedFeature,
+        node.Split.IsOnline(),
+        /*objectSubsetIdx*/ 0, // 0 - learn
+        &indexedSubsetCache,
+        localExecutor,
+        &objectsDataProvider,
+        &columnsIndexing // can return nullptr
+    );
+
+    auto func = BuildNodeSplitFunction(
+        node,
+        *objectsDataProvider,
+        node.Split.Type == ESplitType::OnlineCtr ? &fold.GetCtr(node.Split.Ctr.Projection) : nullptr,
+        /* docOffset */ 0);
+
+    // TODO(ilyzhin) std::function is very slow for calling many times (maybe replace it with lambda)
+    std::function<bool(ui32)> splitFunction;
+    if (!columnsIndexing) {
+        splitFunction = std::move(func);
+    } else {
+        splitFunction = [realObjIdx = columnsIndexing, func=std::move(func)](ui32 idx) {
+            return func(realObjIdx[idx]);
+        };
+    }
+    const size_t blockSize = Max<size_t>(CeilDiv<size_t>(doc_size, localExecutor->GetThreadCount() + 1), 1000);
+    const TSimpleIndexRangesGenerator<size_t> rangesGenerator(TIndexRange<size_t>(doc_size), blockSize);
+    const int blockCount = rangesGenerator.RangesCount();
+    TVector<size_t> n_lefts(blockCount + 1, 0);
+    TVector<size_t> n_rights(blockCount + 1, 0);
+    TVector<std::unique_ptr<ui32> > local_lefts(blockCount);
+    TVector<std::unique_ptr<ui32> > local_rights(blockCount);
+
+    localExecutor->ExecRange(
+        [&node, indicesRef, splitFunction, &docsSubsetPtr, &rangesGenerator, &n_lefts,&n_rights, &local_lefts, &local_rights](int blockId) {
+            const ui32* docsSubsetLocal = docsSubsetPtr;
+            size_t n_left = 0;
+            size_t n_right = 0;
+            const size_t range_size = rangesGenerator.GetRange(blockId).GetSize();
+            ui32* l_ptr = new ui32[range_size];
```

Comment:
we avoid bare pointers as much as possible

Reply:
TVector is used